[docs] Add "Creating a new DebOps role" using Copier dev docs

ypid opened the pull request:
See https://github.com/debops/debops/issues/1617

I guess it could be merged when someone tested this. I probably still should add a hint to the docs, that this is currently just an experiment.

ypid committed:
[docs] Add "Creating a new DebOps role" using Copier dev docs

boucman commented:
Nice, I plan to create y first role pretty soon... I'll definitely play the naive tester for you.

Is the plan to integarte the content of debops-template into the monorepo eventually ?
If yes, maybe the templates could be move to the monorepo in this PR instead of keeping a separate repo ?

boucman commented on docs/developer-guide/new-role.rst:
```diff
@@ -0,0 +1,50 @@
+-----
+
+- Install copier.
+- Clone https://github.com/ypid/debops-template.git next to the debops monorepo (or symlink it next to it).
```
Not sure if cloning is the right way to go here...
The role will probably not be kept in sync with the template repo, it's meant to be it's own thing

maybe give a git-archive command, or point to github's generated tarballs...

ypid replied:
Please read the Copier docs first ;-)

boucman commented on docs/developer-guide/new-role.rst:
```diff
@@ -0,0 +1,50 @@
+
+.. code-block:: shell
+
+   copier -d "author_fullname=$(git config --get user.name)" -d "author_email=$(git config --get user.email)" copy ../../../debops-template/ new_role
```
I assume new_role is the short name of the new role ? this probably needs to be stated explicitely (esp. if there are naming rules...)

ypid replied:
I updated the docs to indicate that this will create a role called `new_role` which should obviously be replaced.

boucman commented on docs/developer-guide/new-role.rst:
```diff
@@ -0,0 +1,50 @@
+Manual post tasks:
+
+- Search for ``TODO`` in the generated role skeleton and address them.
+- Integrate the playbook so that it gets called.
```
How do I do that ? Do I need to modify the monorepo for that ?

ypid replied:
Added a hint.

ypid commented:
> Nice, I plan to create y first role pretty soon... I'll definitely play the naive tester for you.

This makes you the ideal tester than :)

> Is the plan to integarte the content of debops-template into the monorepo eventually ?
If yes, maybe the templates could be move to the monorepo in this PR instead of keeping a separate repo ?

See https://github.com/debops/debops/issues/1617#issue-794558295

boucman commented on docs/developer-guide/new-role.rst:
```diff
@@ -0,0 +1,50 @@
+Create a new rule
+-----------------
+
+To create a role called ``new_role``, in :file:`ansible/roles`, run:
```
Wouldn't it make more sense to have this run from the root of the debops project tree ?

ypid replied:
I thought about that:

Pro root:

* Does not relay on the task trick to move the playbook outside of the generated role scope.

Pro inside role:

* The role directory acts as the "boundary" for Copier where the `.copier-answers.yml` file can be kept. This file cannot be in the root because it should be possible to manage multiple independent roles in the same monorepo. `copier update` needs a path where it can be called.

If you can demonstrate a better way to integrate this into the Monorepo, please do.

boucman replied:
yeah... about that.I was kinda tricked by the copier-answers/yml
In our case, copier is more or less a one-shot thing, you change so much stuff when writing a role that you can't really update using copier

Also, if I want to create mutiple roles, the .copier-answer offers me my previous answers, which is usually not what I want

boucman commented on docs/developer-guide/new-role.rst:
```diff
@@ -0,0 +1,50 @@
+.. code-block:: shell
+
+   copier -d "author_fullname=$(git config --get user.name)" -d "author_email=$(git config --get user.email)" copy ../../../debops-template/ new_role
+
```
There is an assumption here about layout, but I'm not sure what it is

in other words, the previous line (about cloning next to the monorepo) is not clear enough. my monorepo is in ~/.local/share/debops
so I went to ~/tmp, made a symlink to my monorepo, cloned debops-template next to it and now I don't have the right layout.... (i'll figure it out, but the instructions need to be more specific

ypid replied:
Should be fixed now.

boucman reviewed (commented):
Generic remarks 
* Q:Name of project the role manages, in the preferred form.
  That was not clear to me... You assume that the role is about packaging an external project into debops, and that was not my case.
  Once I figured that out, I was good, but that deserves a better explanatory text
* Error on first try
  `mv: impossible de déplacer 'playbook/owadmin.yml' vers '/home/jerros/comite_expertise/adminsys-ecs/ansible/playbooks/service/': N'est pas un dossier` 
  Apparently the directory playbooks/services is not created by debops-init. So you can't assume it's here
* The created playbook imports all sorts of role I didn't need. I'm not sure what the right approch is here... should copier ask ? should they be commented as an example? Not sure... but I had to remove them and nothing told me to review them...
* I'm not sure the license text you generate is ok... Debops does not have a copyright on my code and my code is not part of debops (unless I submit it). It kinda make sense to default to GPLv3+ since that what debops uses so it's probably what the user will need, but other licenses are legal too (especially if the code is not distributed) 


# In the task file
* The role creates all sort of model tasks (cloning git, creating a virtualenv) which are interesting and need to be here somehow, but are not interesting in my case. Removing them was a bit tricky since I needed to figure out what variables to suppress in defaults/main.yml.. Maybe ask what features need to be added one by one in copier ?
* The role seems to rely on various template files (I removed most of them in my case). These were not part of the template.
* I didn't see any `TODO` markers... maybe I missed them ?

ypid commented:
Thanks for your feedback. I hope I addressed it. "The role seems to rely on various template files" will be addressed later.

> Error on first try

Good catch. I was not specific enough in the docs. It needs to be run in the monorepo.

> The created playbook imports all sorts of role I didn't need. I'm not sure what the right approch is here... should copier ask ?

> Maybe ask what features need to be added one by one in copier ?

My take was that deleting is easier than copying together common parts from other roles. One reason for this is that if you copy from other roles, you would have to rename the variables and references from the old role name to the new one. One can forget this (has happened, it is in the git history).

So I left quite a bit of "meat on the bone". It could be made configurable but that I am not sure about maintaining that. Also, what are the common cases? Python stack, LAMP Stack, Ruby on Rails Stack?

> I'm not sure the license text you generate is ok..

Tracked in #1643.

Not sure if we need an option to remove the "This Ansible role is part of DebOps."? The template is specifically for roles that are intended for DebOps so even in "the staging area" the roles could already have to make merging easier.

> The role seems to rely on various template files (I removed most of them in my case).

Good catch. That is work in progress I would say. I guess I will at least add an Ansible local facts skeleton.

> I didn't see any TODO markers... maybe I missed them ?

Seems like it. The generated role has at least one:

```Shell
$ rg TODO
meta/main.yml
32:    - TODO
```

boucman commented:
yes, I see your point about removing is easier than adding, but it's not trivial because you have to figure out what each variable does and which one you can remove.

I'm not sure if there is a way to get the best of both worlds though. Especially since, indeed, adding variables isn't easy either.

boucman commented on docs/developer-guide/new-role.rst:
```diff
@@ -0,0 +1,51 @@
+
+- The generated role has quite a bit of meat on the bone. Just remove the parts you don’t need.
+- Search for ``TODO`` in the generated role skeleton and address them.
+- Integrate the playbook so that it gets called. See how other roles are integrated in :file:`ansible/playbooks/` in the monorepo.
```
That doesn't really help...
AFAICT debops does not have an easy "plug-in" to add external playbooks (I'm not sure how feasible it would be)

I asked on IRC and the answer was to write my own `site.yml` that would be like upstream's one but would import my playbook.

I think it would make more sense if `site.yml` would import an empty `external.yml` playbook that external users could override...

ypid replied:
I also have a top level `site.yml` that includes the debops `site.yml` as one part of it if that helps. See https://github.com/ypid/ypid-ansible-common/blob/master/playbooks/ypid-playbooks/site.yml

What would you rather like to see in the docs? Should I check in Copier tasks if `ansible/playbooks` is available and if not, leave the generated playbook in the role directory so you can do what you want with it when you don’t intent to have the role merged into DebOps directly?

boucman commented on docs/developer-guide/new-role.rst:
```diff
@@ -0,0 +1,51 @@
+- Write the actual role ;-)
+
+Update a role to the latest role standards
+------------------------------------------
```
Roles are so heavily edited, I'm not sure if any form of automatic upgrade makes sense... that would break way too often.

ypid replied:
I had something in the back of my mind when I wrote this. I already excluded heavily edited files from having Copier try to update it using its own mechanism. I agree with you, that would be stupid.

```yaml
_skip_if_exists:
  - defaults/main.yml
  - tasks/main.yml
```

My idea is to use [migrations scripts](https://copier.readthedocs.io/en/latest/configuring/#migrations).
